go/src/lib.rs

```diff
@@ -170,11 +187,166 @@ pub unsafe extern "C" fn close_client(client_adapter_ptr: *const c_void) {
+#[no_mangle]
+pub unsafe extern "C" fn free_command_response(command_response_ptr: *mut CommandResponse) {
+    assert!(!command_response_ptr.is_null());
+    let command_response = unsafe { Box::from_raw(command_response_ptr) };
+    let string_value = command_response.string_value;
+    let string_value_len = command_response.string_value_len;
+    drop(command_response);
+    if !string_value.is_null() {
+        let len = string_value_len as usize;
+        Vec::from_raw_parts(string_value, len, len);
+    }
+}
+
+// TODO: Finish documentation
+/// Converts a double pointer to a vec.
+///
+/// # Safety
+///
+/// * TODO: finish safety section.
+pub unsafe fn convert_double_pointer_to_vec(
+    data: *const *const c_char,
+    len: usize,
+) -> Result<Vec<String>, Utf8Error> {
+    from_raw_parts(data, len)
+        .iter()
+        .map(|arg| CStr::from_ptr(*arg).to_str().map(ToString::to_string))
+        .collect()
+}
+
+pub fn convert_vec_to_pointer<T>(mut vec: Vec<T>) -> (*mut T, c_long) {
+    vec.shrink_to_fit();
+    let vec_ptr = vec.as_mut_ptr();
+    let len = vec.len() as c_long;
+    mem::forget(vec);
+    (vec_ptr, len)
+}
+
+// TODO: Finish documentation
+/// Executes a command.
+///
+/// # Safety
+///
+/// * TODO: finish safety section.
+#[no_mangle]
+pub unsafe extern "C" fn command(
+    client_adapter_ptr: *const c_void,
+    channel: usize,
+    command_type: RequestType,
+    arg_count: usize,
+    args: *const *const c_char,
+) {
+    let client_adapter =
+        unsafe { Box::leak(Box::from_raw(client_adapter_ptr as *mut ClientAdapter)) };
+    // The safety of this needs to be ensured by the calling code. Cannot dispose of the pointer before all operations have completed.
+    let ptr_address = client_adapter_ptr as usize;
+
+    let arg_vec = unsafe { convert_double_pointer_to_vec(args, arg_count) }.unwrap(); // TODO check
+
+    let mut client_clone = client_adapter.client.clone();
+    client_adapter.runtime.spawn(async move {
+        let mut cmd = command_type.get_command().unwrap(); // TODO check cmd
+                                                           //print!("{:?}", cmd.args);
+        cmd.arg(arg_vec);
+
+        let result = client_clone.send_command(&cmd, None).await;
+        let client_adapter = unsafe { Box::leak(Box::from_raw(ptr_address as *mut ClientAdapter)) };
+        let value = match result {
+            Ok(value) => value,
+            Err(err) => {
+                println!(" === err {:?}", err);
+                let message = errors::error_message(&err);
+                let error_type = errors::error_type(&err);
+
+                let c_err_str = CString::into_raw(CString::new(message).unwrap());
+                unsafe { (client_adapter.failure_callback)(channel, c_err_str, error_type) };
+                return;
+            }
+        };
+
+        let mut command_response = CommandResponse {
```

Comment:
Instead, implement the `Default` for `CommandResponse`

Reply:
Done